### DISCUSSION TITLE
Feature Request: required property initialization

### DISCUSSION THREAD

**Discussion by StijnVA** [dotnet/csharplang#678] [General]
# Required Property Initialization

## Summary

Allow an object initializer to invoke a constructor.

## Motivation

Not having this feature leads to confusing mixed object initialisation syntax. It forces the writer of the code to care about the specific initialisation code.

## Detailed design

Consider a class:
```
class MyClass { 
       T1 X {get; set;}
       T2 Y {get; set;}
       T3 Z {get; set;}
      
       MyClass(T1 x, T2 y){
                   X = x;
                   Y = y;
       }
}
```

One could create an object of it:
```
      var o = new MyClass(t1, t2) {
                  Z = t3
       };
```

While it would be nice to write this as:
```
      var o = new MyClass{
                  X = t1,
                  Y = t2,
                  Z = t3
       };
```

## Drawbacks

Having multiple constructors, it might be impossible to differentiate between the constructor used.
Having the class: 

```
class MyClass { 
       T1 X {get; set;}
       T2 Y {get; set;}
       T3 Z {get; set;}
       
       MyClass(){}
      
       MyClass(T1 x, T2 y){
                   X = SomeFunction(x);
                   Y = OtherFunction(y);
       }
}
```

It would be unclear whether the above object construction should translate to:
```
      var o = new MyClass();
       o.X = t1;
       o.Y = t2;
       o.Z = t3;
```
 - or -
```
      var o = new MyClass(t1, t2);
       o.Z = t3;
```

## Alternatives

Having a keyword that requires an property being initialized:

```
class MyClass { 
       T1 X {get; required set;}
       T2 Y {get; required set;}
       T3 Z {get; set;}      
}
```

resulting in a compile error when not specifying a value for X or Y when constructing an object.
```
       var o = new MyClass(); //compile error
       var o = new MyClass { X = t1, Y = t2 }; //this is fine
       var o = new MyClass { X = t1, Y = t2, Z = t3 }; //this is also fine
```

**Comment by GeraudFabien** [comment]
Hello,
You already can do what you want. I forget the name of the feature but you can do

```
class MyClass{
  public int MyProp1 {get; set;}
  public int MyProp2 {get; set;}
}
// Later in some method
var o = new MyClass(){ 
  MyProp1 = 5,
  MyProp2 = 5
 };
```

The only rule is to have public not readonly property. For the "require" part you have enterprise Library and others frameworks who propose thinks. But it's not at compilation time.

**Comment by jnm2** [comment]
@StijnVA That is precisely what constructor parameters are designed for. They enforce every guarantee you need. 

There's no way to enforce guarantees around setting properties at an indeterminate time after the constructor has been called. So many ways to break this...

**Comment by StijnVA** [comment]
Although the above comments are true, i think them are missing the point.

I am just proposing to be able to write:
`var o = new MyClass (x = t1, y = t2) { Z = t3 } `
as
`var o = new MyClass {X = t1, Y = t2, Z = t3 }`
keeping the guaranty that it is impossible to do among other
`var o = new MyClass { X = t1, Z = t3 }`

**Comment by jnm2** [comment]
You can't enforce that guarantee with properties. Imagine that the C# 7.1 compiler enforced it. How would you prevent somebody with a C# 7.0 compiler from compiling `var o = new MyClass { X = t1, Z = t3 }`? What about VB?
